titan.1
.\" Manpage for titan.
.\" Any errors or typos, contact niko@byteptr.com.

.TH man 1 "4 Sep 2018" "1.3" "titan man page"
.SH NAME
titan \- command line password manager
.SH SYNOPSIS
titan [FLAGS] <OPTIONS>
.SH DESCRIPTION
Titan is a command line
password manager and a file encryption program
for Unix like operating systems.
.SH OPTIONS
.IP "-i, --init <path>"
Initialize new database
.IP "-e, --encrypt"
Encrypt current database
.IP "-E, --encrypt-file"
Encrypt a file using strong encryption
.IP "-w  --encrypt-directory"
Encrypt all files in the directory
.IP "-d, --decrypt <path>"
Decrypt database
.IP "-D, --decrypt-file"
Decrypt a file encrypted with Titan
.IP "-W, --decrypt-directory"
Decrypt all files in the directory
.IP "-a, --add"
Add new entry
.IP "-a, --add -I"
Add new entry to database interactively
.IP "-s, --show-db-path"
Show current database path
.IP "-u, --use-db <path>"
Switch using another database
.IP "-r, --remove <id>"
Remove entry pointed by id
.IP "-f, --find <search>"
Search for entries
.IP "-F, --regex <search>"
Search for entries with regular expressions
.IP "-c, --edit <id>"
Edit entry pointed by id
.IP "-f, --find <search>"
Search database
.IP "-l, --list-entry <id>"
List entry pointed by id
.IP "-t, --show-latest <count>"
Show latest entries
.IP "-A, --list-all"
Show database statuses
.IP "-V, --version"
Show program version
.IP "-g, --gen-password <length>"
Generate password
.IP "-q, --quick <search>"
This is the same as running
--show-passwords -f
.IP "-h, --help"
Show short help and exit
.SH FLAGS
.IP "--auto-encrypt"
Automatically encrypt after exit
.IP "--show-passwords"
Show passwords in listings
.IP "--force"
--force only works with --init option
.SH EXAMPLES
Create a new database:
       titan --init "/path/to/file.db"
.PP
Open and decrypt database:
       titan --decrypt "/path/to/existing/file.db"
.PP
Close and encrypt database:
       titan --encrypt

Titan knows what database is currently active and encrypts it.
Encrypting will ask you to type a master passphrase which is used for encryption.
.PP
Add an entry to an open database:
       titan --add
.PP
If you want to export all entries to a file:
       titan --show-passwords -A > file.txt
.PP
To show latest 10 entries:
       titan --show-latest 10
.PP
Encrypting individual files. Since version 1.2 Titan also supports encryption
of individual files. For example to encrypt a file using Titan:
       titan --encrypt-file /home/niko/mysecretdocument.odt
.SH COLORS
Titan supports colored output. To use colors, set an environment variable
TITAN_COLOR with one of the following value:
BLUE, RED, GREEN, YELLOW, MAGENTA, CYAN or WHITE.
.SH NOTES
Titan does not have a concept of "change the master password". When you encrypt
an open database using --encrypt you can type a master password. This password
is then  required to decrypt the database. You can change the master password
everytime when you encrypt the database, if you want to.

Directory encryption is not recursive, this will change in the future versions.
.SH FILES
.I $HOME/.titan.lock
.SH AUTHORS
Written by Niko Rosvall.
.SH COPYRIGHT
Copyright (C) 2018 Niko Rosvall <niko@byteptr.com>
.PP
Released under MIT license.


NEWS
Release 1.3 (2018-09-04)
        *Titan now supports custom count of listing latest entries

Release 1.2 (2017-11-14)
        *Support for colors. See man titan(1) for more information.
        *Added Support encrypting individual files
        *Added support for encrypting directories
        *Many smaller internal changes and bug fixes

titan.c
/*
 * Copyright (C) 2018 Niko Rosvall <niko@byteptr.com>
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <getopt.h>
#include "cmd_ui.h"
#include "entry.h"
#include "db.h"
#include "utils.h"
#include "pwd-gen.h"
#include "crypto.h"

static int show_password = 0;
static int force = 0;
static int auto_encrypt = 0;

static double v = 1.3;

static void version()
{
    printf("Titan version %.1f\n", v);
}

static void usage()
{
#define HELP "\
SYNOPSIS\n\
\n\
    titan [flags] [options]\n\
\n\
OPTIONS\n\
\n\
    -i --init                <path>   Initialize new database\n\
    -e --encrypt                      Encrypt the current password database\n\
    -E --encrypt-file        <path>   Encrypt a file using strong encryption\n\
    -w --encrypt-directory   <path>   Encrypt all files in the directory\n\
    -d --decrypt             <path>   Decrypt password database\n\
    -D --decrypt-file        <path>   Decrypt a file encrypted with Titan\n\
    -W --decrypt-directory   <path>   Decrypt all files in the directory\n\
    -a --add                          Add new entry\n\
    -s --show-db-path                 Show current database path\n\
    -u --use-db              <path>   Switch using another database\n\
    -r --remove              <id>     Remove entry pointed by id\n\
    -f --find                <search> Search entries\n\
    -F --regex               <search> Search entries with regular expressions\n\
    -c --edit                <id>     Edit entry pointed by id\n\
    -l --list-entry          <id>     List entry pointed by id\n\
    -t --show-latest         <count>  Show latest <count> entries\n\
    -A --list-all                     List all entries\n\
    -h --help                         Show short help and exit. This page\n\
    -g --gen-password        <length> Generate password\n\
    -q --quick               <search> This is the same as running\n\
                                      --show-passwords -f\n\
\n\
    -V --version                      Show version number of program\n\
\n\
FLAGS\n\
\n\
    --auto-encrypt                    Automatically encrypt after exit\n\
    --show-passwords                  Show passwords in listings\n\
    --force                           Ignore everything and force operation\n\
                                      --force only works with --init option\n\
\n\
For more information and examples see man titan(1).\n\
\n\
AUTHORS\n\
    Copyright (C) 2018 Niko Rosvall <niko@byteptr.com>\n\
"
    printf(HELP);
}

int main(int argc, char *argv[])
{
    int c;

    if(argc == 1)
    {
        usage();
        return 0;
    }

    while(true)
    {
        static struct option long_options[] =
        {
            {"init",                  required_argument, 0, 'i'},
            {"decrypt",               required_argument, 0, 'd'},
            {"decrypt-file",          required_argument, 0, 'D'},
            {"decrypt-directory",     required_argument, 0, 'W'},
            {"encrypt",               no_argument,       0, 'e'},
            {"encrypt-file",          required_argument, 0, 'E'},
            {"encrypt-directory",     required_argument, 0, 'w'},
            {"add",                   no_argument,       0, 'a'},
            {"remove",                required_argument, 0, 'r'},
            {"find",                  required_argument, 0, 'f'},
            {"regex",                 required_argument, 0, 'F'},
            {"edit",                  required_argument, 0, 'c'},
            {"list-entry",            required_argument, 0, 'l'},
            {"use-db",                required_argument, 0, 'u'},
            {"list-all",              no_argument,       0, 'A'},
            {"help",                  no_argument,       0, 'h'},
            {"version",               no_argument,       0, 'V'},
            {"show-db-path",          no_argument,       0, 's'},
            {"gen-password",          required_argument, 0, 'g'},
            {"quick",                 required_argument, 0, 'q'},
            {"show-latest",           required_argument, 0, 't'},
            {"auto-encrypt",          no_argument,       &auto_encrypt,  1},
            {"show-passwords",        no_argument,       &show_password, 1},
            {"force",                 no_argument,       &force, 1},
            {0, 0, 0, 0}
        };

        int option_index = 0;

        c = getopt_long(argc, argv, "i:d:D:W:eE:w:ar:f:F:c:l:Asu:hVg:q:t:",
                        long_options, &option_index);

        if(c == -1)
            break;

        switch(c)
        {
        case 0:
            /* Handle flags here automatically */
            break;
        case 'i':
            init_database(optarg, force, auto_encrypt);
            break;
        case 'd': //decrypt
            decrypt_database(optarg);
            break;
        case 'D':
            decrypt_any_file(optarg);
            break;
        case 'e': //encrypt
            encrypt_database();
            break;
        case 'E':
            encrypt_any_file(optarg);
            break;
        case 'w':
            encrypt_directory(optarg);
            break;
        case 'W':
            decrypt_directory(optarg);
            break;
        case 'a':
            add_new_entry(auto_encrypt);
            break;
        case 's':
            show_current_db_path();
            break;
        case 'h':
            usage();
            break;
        case 'r':
            remove_entry(atoi(optarg), auto_encrypt);
            break;
        case 'u':
            set_use_db(optarg);
        case 'f':
            find(optarg, show_password, auto_encrypt);
            break;
        case 'F':
            find_regex(optarg, show_password);
            break;
        case 'c':
            edit_entry(atoi(optarg), auto_encrypt);
            break;
        case 'l':
            list_by_id(atoi(optarg), show_password, auto_encrypt);
            break;
        case 'A':
            list_all(show_password, auto_encrypt, -1);
            break;
        case 'V':
            version();
            break;
        case 'g':
            generate_password(atoi(optarg));
            break;
        case 'q':
            show_password = 1;
            find(optarg, show_password, auto_encrypt);
            break;
        case 't':
            show_latest_entries(show_password, auto_encrypt, atoi(optarg));
            break;
        case '?':
            usage();
            break;
        }
    }

    return 0;
}
